Convert to gem

td-extreme opened the pull request:
This pull request converts the TicTacToe game form a command line interface game into a gem file that makes no assumption on the input / output of the game.

The main highlights of what was changed:
- Removed all the classes related to IO as well as the GameLoop class.
- Removed the PlayerHuman class. 
- The AI classes no longer store a player mark.
- The PlayerManager class is no longer responsible for playing turns.
- The GameBoard class now has a play_move method instead of directly accessing the spaces array.
- The TicTacToe class is completely different. It now pieces together the other classes and acts as an API for interacting with those classes.
- There is now a gemspec file for creating the TicTacToe gem.

td-extreme committed:
Removed mark from AI classes. Removed CLI classes.

td-extreme committed:
Removed mark from AI classes.

td-extreme committed:
Updated PlayerManager class.

Changed player manager to have a struct for Player, that stores the mark
and the type. A symbol of :Human is used for a human player. For a
pc_player type is set to a refference of the Ai class to use.

td-extreme committed:
Changed TicTacToe class to be an API.

It is no longer a command line runner for a TicTacToe Game. Instead it
is an API the talks to the other classes.

td-extreme committed:
Removed all the setup logic.

Now that TicTacToe is just an API instead of actually running the game,
this runner no longer needs to build and setup the game.

td-extreme committed:
Removed spec files for classes that are gone.

Had removed several classes earlier, but for some reason the spec files
didn't go away as well.

td-extreme committed:
Fixed test, that was failing.

td-extreme committed:
Update README.md

td-extreme committed:
Updated Test.

td-extreme committed:
Merge branch 'convert_to_gem' of github.com:td-extreme/TicTacToe_Ruby_v2 into convert_to_gem

td-extreme committed:
Updated test again.

td-extreme committed:
Added play_move method to GameBoard, fix ttt test.

td-extreme committed:
Cleaned up code.

td-extreme committed:
Added a gemspec file for making api into a gemfile

td-extreme committed:
Removed spaces from name in gemspec file.

dadarek commented on README.md:
```diff
@@ -1,3 +1,4 @@
 
-[![Build Status](https://travis-ci.org/td-extreme/TicTacToe_Ruby_v2.svg?branch=master)](https://travis-ci.org/td-extreme/TicTacToe_Ruby_v2)
+
+[![Build Status](https://travis-ci.org/td-extreme/TicTacToe_Ruby_v2.svg?branch=convert_to_gem)](https://travis-ci.org/td-extreme/TicTacToe_Ruby_v2)
```
If this gets merged into master, then the master branch will be showing the build status of the `convert_to_gem` branch -- correct?

td-extreme replied:
yeah, I didn't think about that. How would you suggest in the future having a way of showing the current build branch in the README but not having that merge into the master when a pull is made?

dadarek replied:
The typical use case of the build status is to communicate off the home page what's going on with master. If actively someone is working on a feature branch then they're probably just going to CI directly and looking at what's going on there. The build status is sort of a "quick check" to make sure that the master branch on this project is always green (should never ever be red). It's not as evil when other branches temporarily go red, because nobody is (nobody should be) depending on them.

dadarek replied:
So I guess I just would never change this when I'm working on a feature branch. I would be checking CI directly to see what's going on.

td-extreme replied:
okay, that makes sense. Thanks for clarifying that.

dadarek commented on bin/ttt:
```diff
@@ -3,19 +3,5 @@ lib = File.expand_path("../../lib", __FILE__)
-
-myGame = TicTacToe.new(io, p2, p1)
-myGame.play_game
+myGame = TicTacToe.new
```
If the `bin/ttt` executable remains in the gem, how would clients of the gem use it?

td-extreme replied:
The bin/ttt file should probably get removed. Right? I didn't include it in the list of files in the gemspec. Also, since you can no longer run this as a stand alone program it doesn't make sense to have the bin/ttt file.

dadarek replied:
If it servers no purpose - dump it.

dadarek replied:
And that goes for every file and line of code in a project.

dadarek commented on lib/player_manager.rb:
```diff
@@ -1,23 +1,31 @@
-require 'game_state'
+require 'ai_basic'
```
I don't see `AiBasic` or `AiMinimax` being used anywhere in this class - are the requires required?

td-extreme replied:
You're right they are not used and I could and should removed those two requires.

dadarek commented on lib/player_manager.rb:
```diff
@@ -1,23 +1,31 @@
+  Player = Struct.new(:mark, :type)
+
+  def initialize()
+      @player1 = Player.new(:X, :Human)
```
What is the reasoning behind always initializing both players to be `:Human` ?

dadarek replied:
On an unrelated note - symbols typically won't begin with an upper case.  snake_case is what you'll usually see.

td-extreme replied:
The way I the type works is that it either points to a symbol `:human` or to one of the Ai classes. It just seemed easier to have both initialize as human as default then have the driver program using the gem change decided if one should be a PC and what skill level it should be.

dadarek commented on lib/tictactoe.rb:
```diff
@@ -1,42 +1,78 @@
+  end
+
+  def get_ai_player_move
+    @players.current_player.type.get_move(get_game_state) if is_current_player_ai?
```
The default `type` of players initialized in `PlayerManager` is the symbol `:Human`. If a client doesn't change that after construction of a `PlayerManager`, will this line blow up here?

td-extreme replied:
It should just return nil if the player is human. Let me make sure this won't have any unintended results else where.

dadarek replied:
Ah - I didn't see the `if` guard

dadarek commented on lib/tictactoe.rb:
```diff
@@ -1,42 +1,78 @@
+    @players.current_player.type.respond_to?(:get_move)
+  end
+
+  def get_game_state
```
What's the reason for always creating a new game state?

Also - if that is necessary - does make sense to call this method `create` or something else that is a little more explicit about this fact?

td-extreme replied:
The game state is used by the AI so that it can play out moves on a copy of the board and not worry about having to reseting the original board.

dadarek replied:
Roger

dadarek commented on tictactoe.gemspec:
```diff
@@ -0,0 +1,13 @@
+
```
New line?

td-extreme replied:
woops!!

dadarek commented on tictactoe.gemspec:
```diff
@@ -0,0 +1,13 @@
+  s.description = "A Tic Tac Toe gem"
+  s.authors     = ["Tyler Decker"]
+  s.email       = 'tyler.decker@gmail.com'
+  s.files       = ["lib/ai_basic.rb", "lib/ai_minimax.rb", "lib/game_board.rb", "lib/game_rules.rb", "lib/game_state.rb", "lib/player_manager.rb", "lib/tictactoe.rb"]
```
Try something like:
  `s.files = Dir["app/**/*"] + ["Rakefile", "README"]`

dadarek commented on spec/tictactoe_spec.rb:
```diff
@@ -1,31 +1,131 @@
+  let(:players) { double }
+  let(:state) { double }
+  let(:ai_basic) { AiBasic.new(GameRules.new) }
+  Player = Struct.new(:mark, :type)
```
You already define this in a lib file somewhere .. does it make sense to expose it so you can use it here? Or is that strictly internal?

If it does _not_ make sense to expose it and you just need something that response to these two methods, an rspec `double(method_1: :value, method_2: :value_2)` is sufficient.

There is also `OpenStruct` which you can use super-duper dynamically (scary, but convenient, which usually means it'll bite you later). But it's there if you're feeling adventurous - google it.

dadarek commented on spec/player_manager_spec.rb:
```diff
@@ -1,41 +1,42 @@
-      end
+
+  describe "Setting Player's mark and type" do
+    it "Default mark for player1 is :X" do
```
What are your thoughts on libraries that set up a bunch of defaults for you v.s. a libraries that make you explicitly define the things that are relevant?

td-extreme replied:
My thoughts with having defaults was that it makes it so if the program using the gem forgets to set something it still works. But having to explicitly set things seems to make more sense, as that way it guarantees the program didn't forget to set something.

dadarek commented:
What was your reasoning for choosing to go down the route of "let's make this a gem, publish it, and then use it in an application somewhere else", v.s. "let's create an application and see what it needs and then mold this gem so that it provides what the application needs"?

Have you considered other approaches to tackling this task?

mikeknep commented on lib/tictactoe.rb:
```diff
@@ -1,42 +1,78 @@
+class TicTacToe
 
-  def initialize (game_io, player1, player2)
+  def initialize
```
This change severely impacts the testability of this class—instead of passing in dependencies they are now hardcoded. I don't think that should be necessary. What I _suspect_ was going on here (correct me if I'm wrong) is you were thinking any application that ends up using this gem shouldn't need to know too many details about how to wire things up together. There is some truth to that, but I think you may have gone a little overboard. Turning the core TTT logic into a gem shouldn't require abandoning all the nice design patterns you originally implemented when it was a full application.

Thoughts?

td-extreme replied:
Mike, 
Yes, my thought was that the user of the gem should just be able to create a TicTacToe game and not have to wire parts together. However, the dependencies passed in that were removed didn't seem to apply as much to the game if it was no longer a terminal application. IE the game_io should be handled by the application using the gem. Also I removed the player classes as the symbols used for players would no longer need to be changeable. I see no reason to have the internal board change what symbols are used for player1 and player2. My thought was that a web app or any other application using this gem would decide for its self how to display the board. ie if the value of the space is :X the web app would say okay put the picture or character I use for :X in that spot.

td-extreme replied:
Darek, 
I could have started by making a web app or terminal app that just used a mock of the TicTacToe gem. Then after completing that I could have then wrote the gem file. 

I decided to go with the bottom up approach because I already had the tictactoe code written. Also, I had never made a gem before so I kind of wanted to attempt doing that. I did notice that, I was spending time trying to figure out what methods I would need to provide for an application that didn't exist yet. This also made me worry that I might be forgetting something, or that I was writing methods that would never actually be needed. 

I think what I would like to do moving forward, is to forget about this branch, try a top down approach and work on the web version and terminal version of the game, and after completing those create a separate branch off of the master branch and create the gem again with the knowledge of knowing what methods a program using it would expect to have available.

td-extreme committed:
Updated PlayerManager to take 2 arguments.

This arguments set the type for each player.

td-extreme committed:
Updated TicTacToe

It now takes two arguments for its constructor and passes those to the
PlayerManager. Removed the methods to set player types. changed
play_human_move to play_move.

td-extreme committed:
Rebuilt GemFile

td-extreme committed:
Update README.md

dadarek commented:
Did you intend to merge this?
